File: classes/spotify.py
# https://developer.spotify.com/documentation/web-api

from .abstract_platform import Platform
import re
import os
import requests
import base64
import time
from dotenv import load_dotenv

load_dotenv(os.path.join(os.path.dirname(os.path.dirname(__file__)), ".env"))

class Spotify(Platform):
    client_id = os.environ.get("SPOTIFY_CLIENT_ID")
    client_secret = os.environ.get("SPOTIFY_CLIENT_SECRET")
    access_token = ""
    token_type = ""
    token_expire_in = None
    startTime = None
    headers = {"Authorization": f"Bearer {access_token}"}

    def is_token_valid(self):
        if not self.access_token or not self.startTime or not self.token_expire_in:
            return False
        return (time.time() - self.startTime) < self.token_expire_in        

    def __init__(self, url):
        super().__init__(url)
        self.track_id = None
        if url:
            match = re.search(r"/track/([a-zA-Z0-9]+)", url)
            self.track_id = match.group(1) if match else None
        if not self.is_token_valid():
            self.get_spotify_token()

    def get_spotify_token(self):
        auth_str = f"{self.client_id}:{self.client_secret}"
        b64_auth_str = base64.b64encode(auth_str.encode()).decode()

        headers = {
            "Authorization": f"Basic {b64_auth_str}",
            "Content-Type": "application/x-www-form-urlencoded"
        }
        data = {"grant_type": "client_credentials"}

        response = requests.post("https://accounts.spotify.com/api/token", headers=headers, data=data)
        token_info = response.json()

        self.access_token = token_info.get("access_token")
        self.token_type = token_info.get("token_type")
        self.token_expire_in = token_info.get("expires_in")
        self.headers = {"Authorization": f"Bearer {self.access_token}"}
        self.startTime = time.time()

    def get_title(self):
        response = requests.get(f"https://api.spotify.com/v1/tracks/{self.track_id}", headers=self.headers)
        data = response.json()
        return data["name"]

    def get_artist(self):
        response = requests.get(f"https://api.spotify.com/v1/tracks/{self.track_id}", headers=self.headers)
        data = response.json()
        return data["artists"][0]["name"]

    def get_link(self):
        return self.url
    
    def search_link(self, track_name, track_artist):
        query = f'track:"{track_name}" artist:"{track_artist}"'
        url = "https://api.spotify.com/v1/search"
        params = {
            "q": query,
            "type": "track",
            "limit": 1
        }
        headers = {"Authorization": f"Bearer {self.access_token}"}
        response = requests.get(url, headers=headers, params=params)
        if response.status_code == 200:
            data = response.json()
            items = data.get("tracks", {}).get("items", [])
            if items:
                return items[0]["external_urls"]["spotify"]
        return None

File: src/platforms/spotify.rs
use base64::{Engine as _, engine::general_purpose};
use once_cell::sync::Lazy;
use reqwest;
use serde::Deserialize;
use std::env;
use std::sync::Mutex;
use std::time::{Duration, Instant};

use crate::Error;
use crate::types::TrackInfo;

static SPOTIFY_TOKEN: Lazy<Mutex<Option<SpotifyToken>>> = Lazy::new(|| Mutex::new(None));

#[derive(Deserialize, Debug)]
struct SpotifyTokenResponse {
    access_token: String,
    expires_in: u64,
}

#[derive(Debug)]
struct SpotifyToken {
    access_token: String,
    expires_in: u64,
    created_at: Instant,
}

#[derive(Deserialize, Debug)]
struct SpotifyTrack {
    name: String,
    artists: Vec<SpotifyArtists>,
    album: SpotifyAlbum,
}

#[derive(Deserialize, Debug)]
struct SpotifyArtists {
    name: String,
}

#[derive(Deserialize, Debug)]
struct SpotifyAlbum {
    name: String,
}

#[derive(Deserialize, Debug)]
struct SpotifySearchTrack {
    external_urls: SpotifyExternalUrls,
}

#[derive(Deserialize, Debug)]
struct SpotifyExternalUrls {
    spotify: String,
}

#[derive(Deserialize, Debug)]
struct SpotifySearchTracks {
    items: Vec<SpotifySearchTrack>,
}

#[derive(Deserialize, Debug)]
struct SpotifySearchResponse {
    tracks: SpotifySearchTracks,
}

fn extract_track_id(url: &str) -> Option<String> {
    let track_part = url.split("/track/").nth(1)?;
    let track_id = track_part.split('?').next()?.split('/').next()?.to_string();

    if !track_id.is_empty() {
        Some(track_id)
    } else {
        None
    }
}

impl SpotifyToken {
    fn is_valid(&self) -> bool {
        self.created_at.elapsed() < Duration::from_secs(self.expires_in)
    }
}

async fn get_access_token() -> Result<SpotifyToken, Error> {
    let client_id = env::var("SPOTIFY_CLIENT_ID")?;
    let client_secret = env::var("SPOTIFY_CLIENT_SECRET")?;
    let auth = format!("{}:{}", client_id, client_secret);
    let b64_auth = general_purpose::STANDARD.encode(auth);

    let client = reqwest::Client::new();
    
    let res = client
        .post("https://accounts.spotify.com/api/token")
        .header("Authorization", format!("Basic {}", b64_auth))
        .header("Content-Type", "application/x-www-form-urlencoded")
        .body("grant_type=client_credentials")
        .send()
        .await
        .map_err(|e| format!("Spotify token request failed: {}", e))?;

    let token: SpotifyTokenResponse = res.json().await?;
    Ok(SpotifyToken {
        access_token: token.access_token,
        expires_in: token.expires_in,
        created_at: Instant::now(),
    })
}

async fn get_access_token_cached() -> Result<String, Error> {
    {
        let token_guard = SPOTIFY_TOKEN.lock().unwrap();
        if let Some(token) = token_guard.as_ref() {
            if token.is_valid() {
                return Ok(token.access_token.clone());
            }
        }
    }
    let new_token = get_access_token().await?;
    let access_token = new_token.access_token.clone();
    let mut token_guard = SPOTIFY_TOKEN.lock().unwrap();
    *token_guard = Some(new_token);
    Ok(access_token)
}

pub async fn get_track_info(_url: &str) -> Result<TrackInfo, Error> {
    let track_id = extract_track_id(_url).ok_or("Invalid Spotify track URL")?;
    let access_token = get_access_token_cached().await?;
    let client = reqwest::Client::new();

    let url = format!("https://api.spotify.com/v1/tracks/{}", track_id);
    let res = client.get(&url).bearer_auth(&access_token).send().await?;

    let track: SpotifyTrack = res.json().await?;
    let artists = track
        .artists
        .iter()
        .map(|a| a.name.as_str())
        .collect::<Vec<_>>()
        .join(", ");

    Ok(TrackInfo {
        original_platform: "Spotify".to_string(),
        title: track.name,
        artist: artists,
        album: Some(track.album.name),
    })
}

pub async fn get_track_link(info: &TrackInfo) -> Option<String> {
    let access_token = get_access_token_cached().await.ok()?;

    let url = format!(
        "https://api.spotify.com/v1/search?q=track:\"{}\" artist:\"{}\"&type=track&limit=1",
        info.title, info.artist
    );

    let client = reqwest::Client::new();
    let res = client
        .get(&url)
        .bearer_auth(&access_token)
        .send()
        .await
        .ok()?;

    let search_result: SpotifySearchResponse = res.json().await.ok()?;

    search_result
        .tracks
        .items
        .first()
        .map(|track| track.external_urls.spotify.clone())
}
